src/protocols/bdx/BDXMessageUtils.h

```diff
@@ -0,0 +1,204 @@
+    bool mDefLen;          ///< True if transfer has a definite length.
+
+    uint64_t mMaxLength;    ///< Proposed max length of data in transfer, 0 for indefinite (required if mDefLen is set)
+    uint16_t mMaxBlockSize; ///< Proposed max block size to use in transfer (required)
+
+    // File designator (required)
+    uint8_t * mFileDesignator;
+    uint16_t mFileDesLength; ///< Length of file designator string (not including null-terminator)
+
+    // Additional metadata (optional, TLV format)
+    uint8_t * mMetadata;
+    uint16_t mMetadataLength;
+};
+
+struct SendInit : public TransferInit
+{
+};
+
+struct ReceiveInit : public TransferInit
+{
+};
+
+/*
+ * A structure for representing a SendAccept message.
+ */
+struct SendAccept
+{
+public:
+    /**
+     * @brief
+     *  Pack (write) the message into a PacketBuffer.
+     *
+     * @param[out] aBuffer a PacketBuffer to use to write the message
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to write to the PacketBuffer
+     */
+    CHIP_ERROR Pack(System::PacketBuffer * aBuffer);
+
+    /**
+     * @brief
+     *  Parse data from an PacketBuffer into a struct instance
+     *
+     * @param[in] aBuffer Pointer to a PacketBuffer containing the data
+     * @param[out] aParsedMessage Reference to a struct instance where results will be stored
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to read the message
+     */
+    static CHIP_ERROR Parse(System::PacketBuffer * aBuffer, SendAccept & aResponse);
+
+    /**
+     * @brief
+     *  Equality check method.
+     */
+    bool operator==(const SendAccept &) const;
+
+    // All required
+    uint8_t mVersion;         ///< The agreed upon version for the transfer.
+    ControlMode mControlMode; ///< Agreed upon transfer control method
+    uint16_t mMaxBlockSize;   ///< Chosen max block size to use in transfer
+
+    // Additional metadata (optional, TLV format)
+    uint8_t * mMetadata;
+    uint16_t mMetadataLength;
+};
+
+/**
+ * @class ReceiveAccept
+ *
+ * @brief
+ *   The ReceiveAccept message is used to accept a proposed exchange when the
+ *   receiver is the initiator.
+ */
+struct ReceiveAccept
+{
+public:
+    /**
+     * @brief
+     *  Pack (write) the message into a PacketBuffer.
+     *
+     * @param[out] aBuffer a PacketBuffer to use to write the message
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to write to the PacketBuffer
+     */
+    CHIP_ERROR Pack(System::PacketBuffer * aBuffer);
+
+    /**
+     * @brief
+     *  Parse data from an PacketBuffer into a struct instance
+     *
+     * @param[in] aBuffer Pointer to a PacketBuffer containing the data
+     * @param[out] aParsedMessage Reference to a struct instance where results will be stored
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to read the message
+     */
+    static CHIP_ERROR Parse(System::PacketBuffer * aBuffer, ReceiveAccept & aResponse);
```

Comment:
would this look better as a constructor that takes a packetbuffer?

Reply:
That's a great idea, but what if there's an error reading from the packetbuffer? How would we convey that?

Reply:
You could have a `CHIP_ERROR *` outparam....  A little ugly, but might be ok.

Reply:
@holbrookt: good point, something I didn't consider

Reply:
My gut is telling me that this works better as a static method than a constructor, due to the amount of logic involved and needing to return an error. Is that ok with you?

I think a constructor would be better if we wanted to make the members private, but my intention was to have the fields be public to make it easier to fill out an outgoing message or read from an incoming message.

Reply:
Oh sorry, just realized you marked this as resolved 👍